### PR TITLE
HAAR-1385: cache Delius roles

### PR DESCRIPTION
**Pull request by ravmoj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/manage/service/ManageUsersApiService.kt [hunk 1]
```diff
@@ -27,16 +29,34 @@ class ManageUsersApiService(
     private val log = LoggerFactory.getLogger(this::class.java)
   }
 
-  fun getDeliusMappedRoles(userName: String, roles: String): List<UserRole> {
+  private lateinit var existingDeliusRoles: LoadingCache<String, Map<String, List<String>>>
+
+  private var userName: String = "hmpps-auth"
+  init {
+    existingDeliusRoles = CacheBuilder.newBuilder().expireAfterWrite(1, TimeUnit.DAYS).build(
+      CacheLoader.from { _: String ->
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/manage/service/ManageUsersApiService.kt:37]: Just wondering that in the future we might want this to be an env var - is fine for now though.  Like the cachebuilder.
Should we be using imports for the statics?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/manage/service/ManageUsersApiService.kt [hunk 2]
```diff
@@ -27,16 +29,34 @@ class ManageUsersApiService(
     private val log = LoggerFactory.getLogger(this::class.java)
   }
 
-  fun getDeliusMappedRoles(userName: String, roles: String): List<UserRole> {
+  private lateinit var existingDeliusRoles: LoadingCache<String, Map<String, List<String>>>
+
+  private var userName: String = "hmpps-auth"
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/manage/service/ManageUsersApiService.kt:34]: Not totally sure why the username is required

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/manage/service/ManageUsersApiService.kt [hunk 3]
```diff
@@ -27,16 +29,34 @@ class ManageUsersApiService(
...
+
+  fun getDeliusRolesFromCache(userName: String): LoadingCache<String, Map<String, List<String>>> {
+    this.userName = userName
+    if (existingDeliusRoles.get("DELIUS_ROLES").isEmpty()) {
+      refreshDeliusRolesCache()
+    }
+    return existingDeliusRoles
+  }
+
+  fun getDeliusMappedRoles(userName: String): Map<String, List<String>> {
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/manage/service/ManageUsersApiService.kt:51]: Do we have any tests for this (maybe I've missed them)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/manage/service/ManageUsersApiService.kt (hunk 1 above)
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/manage/service/ManageUsersApiService.kt]: added static imports

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/manage/service/ManageUsersApiService.kt (hunk 2 above)
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/manage/service/ManageUsersApiService.kt:34]: username is part of error message as followed in other clients (nomis, delius) in auth service

**Review by karenmillermoj — APPROVED** [review]
lgtm

**Pull request merged**